src/hotspot/share/jeandle/jeandleAbstractInterpreter.cpp

```diff
@@ -139,46 +139,47 @@ bool JeandleVMState::update_phi_nodes(JeandleVMState* income_jvm, llvm::BasicBlo
 void JeandleVMState::push(BasicType type, llvm::Value* value) {
   assert(value != nullptr, "null value to push");
   assert(value->getType() == JeandleType::java2llvm(type, *_context), "type must match");
-  _stack.push_back(value);
+  _stack.push_back(TypedValue(type, value));
   if (is_double_word_type(type)) {
-    _stack.push_back(nullptr);
+    _stack.push_back(TypedValue::null_value());
   }
 }
 
 llvm::Value* JeandleVMState::pop(BasicType type) {
   if (is_double_word_type(type)) {
-    assert(_stack.back() == nullptr, "hi-word of doubleword value must be null");
+    assert(_stack.back().is_null(), "hi-word of doubleword value must be null");
     _stack.pop_back();
   }
-  llvm::Value* v = _stack.back();
-  assert(v != nullptr, "null value to pop");
-  assert(v->getType() == JeandleType::java2llvm(type, *_context), "type must match");
+  TypedValue v = _stack.back();
+  assert(v.value() != nullptr, "null value to pop");
+  assert(v.computational_type() == JeandleType::actual2computational(type), "type must match");
   _stack.pop_back();
-  return v;
+  return v.value();
 }
 
 // Locals operations:
 
 llvm::Value* JeandleVMState::load(BasicType type, int index) {
-  assert(!is_double_word_type(type) || _locals[index + 1] == nullptr, "hi-word of doubleword value must be null");
-  assert(_locals[index] != nullptr, "null value to load");
-  assert(_locals[index]->getType() == JeandleType::java2llvm(type, *_context), "type must match");
-  return _locals[index];
+  assert(!is_double_word_type(type) || _locals[index + 1].is_null(), "hi-word of doubleword value must be null");
+  TypedValue v = _locals[index];
+  assert(v.value() != nullptr, "null value to load");
+  assert(v.computational_type() == JeandleType::actual2computational(type), "type must match");
```

Comment:
Here we introduce an implicit constraint that `JeandleType::java2llvm(type) == JeandleType::java2llvm(JeandleType::actual2computational(type))`.
So I think it's better to re-implement `JeandleType::java2llvm` using `JeandleType::actual2computational`, and add more comment to explain this constraint and why we have this constraint.

Reply:
Now computational type is only used for checking. Do we need use computational type here?